File: brainfuck.c
#include<stdio.h>
#include<stdlib.h>
#include<string.h>

#define CODE_LENGTH 4096
#define STACK_FRAME 1024
#define DATA_LENGTH 4096
//code area
char code[CODE_LENGTH];
//stack frame (for nested loop)
char* stack[STACK_FRAME];
//runtime data
char data[DATA_LENGTH];
//code pointer
char* cp=code;
//stack pointer
char** sp=stack;
//data pointer
char* dp=data;
/*
 * function mapping, inspired by a blog.
 * key:instruction character, ASCII
 * value:instruction function, function pointer
 */
typedef void (*INS)();
/*
 * according to ASCII table, ']' is the last character of brainfuck's
 * reserved word and its value is 93, so, to save memory, 94 is enough.
 */
INS ins[94];

/*
 * right shift
 */
void rShift(){
    dp++;
}
/*
 * left shift
 */
void lShift(){
    dp--;
}
/*
 * increase one
 */
void increase(){
    (*dp)++;
}
/*
 * decrease one
 */
void decrease(){
    (*dp)--;
}
/*
 * input
 */
void input(){
    *dp=getchar();
}
/*
 * output
 */
void output(){
    putchar(*dp);
}
/*
 * begin loop
 */
void begin(){
	if(*dp){
		//save loop start position to stack frame
		*sp++=cp-1;
	}
	else{
		//skip to next related ]
		int nested=1;
		for(cp++;*cp&&nested;cp++){
			if(*cp=='[') nested++;
			if(*cp==']') nested--;
		}
	}
}
/*
 *end loop
 */
void end(){
	if(*dp) cp=*(--sp);
}

int main(int argc,char** argv){
	if(argc!=2){
		printf("Usage:%s sourceFile\n",argv[0]);
	}
	else{
		freopen(argv[1], "r", stdin);
		char c;
		while((c = getchar())!= EOF){
			if(strchr("<>+-[].,",c))
				*cp++=c;
			//ignore whitespace and unvalid character
			else if(strchr(" \n\r\t",c))
				continue;
			else
				printf("ignore unvalid character:%c\n",c);
		}
		//reset
		cp=code;
		ins['>']=rShift;
		ins['<']=lShift;
		ins['+']=increase;
		ins['-']=decrease;
		ins[',']=input;
		ins['.']=output;
		ins['[']=begin;
		ins[']']=end;
		while(*cp) ins[*cp++]();
	}
}

File: brainfuck.java
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class Brainfuck{
	interface Instruction{
		void execute();
	}
	private static final int CODE_LENGTH=4096;
	private static final int STACK_FRAME=1024;
	private static final int DATA_LENGTH=4096;
	
	private static final String reserved="<>+-[].,";
	private static final String whitespace=" \n\r\t";
	
	private static final byte[] code=new byte[CODE_LENGTH];
	private static final int[] stack=new int[STACK_FRAME];
	private static final byte[] data=new byte[DATA_LENGTH];
	
	private static int cp,sp,dp;
	private static Map<Byte, Instruction> ins=new HashMap<>();
	private static Scanner in=new Scanner(System.in);
	
	static{
		ins.put((byte)'>',()->dp++);
		ins.put((byte)'<',()->dp--);
		ins.put((byte)'+',()->data[dp]++);
		ins.put((byte)'-',()->data[dp]--);
		ins.put((byte)',',()->data[dp]=in.nextByte());
		ins.put((byte)'.',()->System.out.print((char)data[dp]));
		ins.put((byte)'[',()->{
			if(data[dp]!=0){
				stack[sp++]=cp-1;
			}
			else{
				int nested=1;
				for(cp++;code[cp]!=0&&nested!=0;cp++){
					if(code[cp]=='[') nested++;
					if(code[cp]==']') nested--;
				}
			}
		});
		ins.put((byte)']',()->{
			--sp;
			if(data[dp]!=0){
				cp=stack[sp];
			}
		});
	}
	public static void execute(String file) throws IOException{
		String content=new String(Files.readAllBytes(Paths.get(file)),Charset.forName("UTF-8"));
		for(char c:content.toCharArray()){
			if(reserved.contains(String.valueOf(c)))
				code[cp++]=(byte)c;
			else if(whitespace.contains(String.valueOf(c)))
				continue;
			else
				System.out.println("ignore unvalid character:"+c);
		}
		cp=0;
		while(code[cp]!=0) ins.get(code[cp++]).execute();
	}
	public static void main(String[] args){
		if(args.length!=1){
			System.out.println("Usage:Brainfuck sourceFile");
		}
		else{
			try {
				execute(args[0]);
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	}
}